### PR TITLE
Add kc_action to redirect URI after a required action is cancelled (#31894)

### LINKED ISSUE #31894: Redirect after cancelling a required action should contain kc_action parameter
### Description

The cancelled required action should be added as a redirect parameter `kc_action` to the redirect URL.

### Discussion

_No response_

### Motivation

Clients that trigger an initiate application initiated actions should be able to detect
which action was cancelled when the browser is redirected back to the application.
Currently client applications need to use local state for storing a reference to the triggered
action.

### Details

Current redirect after `UPDATE_PASSWORD` required action was cancelled:

```
http://localhost:8081/auth/realms/demo/account/account-security/signing-in?state=d50b802d-5f35-4cec-9c6e-c365d35a1e83&session_state=cc2c37aa-c8e6-4df0-a08d-9a87de5a74bf&iss=http%3A%2F%2Flocalhost%3A8081%2Fauth%2Frealms%2Fdemo&kc_action_status=cancelled&code=a8b88e72-ce91-49ab-9992-6e129ef7a56d.cc2c37aa-c8e6-4df0-a08d-9a87de5a74bf.6ae0ea88-01dd-47c7-9405-09c664efc4a0
```

Note that the name of the required action is not present in the redirect URL.

The last executed action can be added for OIDC in the following place:
- [org.keycloak.protocol.oidc.OIDCLoginProtocol#authenticated](https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/protocol/oidc/OIDCLoginProtocol.java#L232)

Old:
```java
if (kcActionStatus != null) {
      redirectUri.addParam(Constants.KC_ACTION_STATUS, kcActionStatus);
 }
```

New:
```java
if (kcActionStatus != null) {
      String requiredActionAlias = authSession.getAuthNote(AuthenticationProcessor.LAST_PROCESSED_EXECUTION);
      if (requiredActionAlias != null) {
          redirectUri.addParam(Constants.KC_ACTION, requiredActionAlias);
      }
      redirectUri.addParam(Constants.KC_ACTION_STATUS, kcActionStatus);
 }
```

With that change the required action is passed in the redirect URL.
```
http://localhost:8081/auth/realms/demo/account/account-security/signing-in?state=d4fa0136-f589-4f54-916e-771a388085bf&session_state=cc2c37aa-c8e6-4df0-a08d-9a87de5a74bf&iss=http%3A%2F%2Flocalhost%3A8081%2Fauth%2Frealms%2Fdemo&kc_action=UPDATE_PASSWORD&kc_action_status=cancelled&code=59af874e-aaea-47f8-be8f-e6d0a0f6405b.cc2c37aa-c8e6-4df0-a08d-9a87de5a74bf.6ae0ea88-01dd-47c7-9405-09c664efc4a0
```

### PR DESCRIPTION
**Pull request by thomasdarimont**
Previously, the kc_action parameter was missing and clients that wanted to know which action got cancelled needed to remember which action was executed.

We now pass the last executed required action, which was cancelled as the kc_action parameter in the redirect URI such that clients can know about which action was cancelled by the user.

I also fixed the order of parameters in ` assertThat(kcActionStatus, is(expectedStatus));` which was wrong.

Fixes #31894

### CODE REVIEW

**Comment by mposolda** [comment]
@thomasdarimont I think this is great addition, Thanks Thomas!

Do you think it is possible to add some short notice about this into the AIA documentation section https://www.keycloak.org/docs/latest/server_admin/index.html#con-aia_server_administration_guide ?

Also I think that some of the failing tests are caused by the change in the behaviour caused by this PR. I am pretty sure that WebAuthn tests are failing because of the `kc_action` now present in the redirect URI. But the `BruteForceTest` failures are probably unrelated, so you can likely ignore those (There is WiP to fix those).

**Comment by thomasdarimont** [comment]
Thanks @mposolda I rebased the PR and added a small section to the aia docs.

**Comment by mposolda** [comment]
@thomasdarimont Thanks! Still there are the failures in "WebAuthn IT" tests, which look like a regression of this PR.

**Comment by thomasdarimont** [comment]
Thanks Marek I take a look at this later

**Comment by thomasdarimont** [comment]
@mposolda I gave this another look. 

I struggled to fix this for a while, but I found a solution now.

Anlong the way I found a few issues:
1) Too strict URL checking in AbstractPage#isCurrent implementations.
The implementation of `org.keycloak.testsuite.page.AbstractPage#isCurrent` used by all pages in the e2e tests use is quite strict and tests the whole URI string **including** parameters, however in order to determine the current page checking the http(s)://host:port/path components should be enough. This strict checking breaks if there are any new query parameters.
I think we should provide a URL check that only checks for the baseUrl + page path.

2) If I emit the kc_action parameter alongside kc_action_status=cancelled in the redirect URI in `org.keycloak.protocol.oidc.OIDCLoginProtocol#authenticated`, then (at least in the tests) the kc_action parameter is added to the resulting page from the account-console application, which fails the test due to the strict ìsCurrent()` check mentioned in 1).

To fix 2) I needed to make keycloak.js aware of the `kc_action` query parameter in the redirect as well, and add an `action` parameter to the onActionUpdate(..) callback function definition.
With this clients using keycloak.js can now detect which action had the outcome of cancelled/successful/error.

The tests should work now, however users that do not update their keycloak.js library might see the issues from 2). I think we should mention this in the migration guide.

**Comment by mposolda** [comment]
@thomasdarimont Thanks for the investigation Thomas! I am approving and merging. If you please have a chance to send a small separate PR for the upgrading guide, it can be great!

For the `AbstractPage.isCurrent()` , it is strict as you mentioned. On the other hand, it helped to show that the update in `keycloak.js` is needed for this change as well, which would otherwise be left undetected :-) So maybe being more strict is not so bad though, not sure...

**Review by mposolda — DISMISSED** [review]

**Comment by mposolda** [comment]
@thomasdarimont Cannot merge yet as this needs review from UI team as well.

**Comment by thomasdarimont** [comment]
thanks @mposolda! I added a small note to the upgrading / changes doc

**Review by mposolda — DISMISSED** [review]
@thomasdarimont Thanks!

@keycloak/ui-maintainers Anyone available to review this PR please?

File: docs/documentation/upgrading/topics/changes/changes-26_0_0.adoc
Comment (jonkoops) [inline docs/documentation/upgrading/topics/changes/changes-26_0_0.adoc:133]:
```suggestion
Note: This feature required changes to the Keycloak JS adapter, therefore it is recommended to upgrade to the latest version of the adapter if you want to make use of this feature.
```

File: js/libs/keycloak-js/dist/keycloak.d.ts
Comment (jonkoops) [inline js/libs/keycloak-js/dist/keycloak.d.ts:537]: Is `action` here a finite amount of options that are static in nature? If so can we list all the options explicitly as we do with `status` before it? If not, this comment can be ignored.

File: js/libs/keycloak-js/dist/keycloak.d.ts
Comment (jonkoops) [inline js/libs/keycloak-js/dist/keycloak.d.ts:535]: Thanks for adding some docs here, I feel though we can be a bit more explicit and either link to documentation that explains what the possible options are here. Or even better, document these values in-line.

**Review by jonkoops — CHANGES_REQUESTED** [review]

File: docs/documentation/upgrading/topics/changes/changes-26_0_0.adoc
Comment (thomasdarimont) [reply docs/documentation/upgrading/topics/changes/changes-26_0_0.adoc:133]: Thanks I'll apply your suggestion.

File: js/libs/keycloak-js/dist/keycloak.d.ts
Comment (thomasdarimont) [reply js/libs/keycloak-js/dist/keycloak.d.ts:537]: The values for the `action` parameter are the provider aliases for required actions like `UPDATE_PASSWORD`, `CONFIGURE_TOTP`  etc. Some of those required action aliases are indeed static values (see: [org.keycloak.models.UserModel.RequiredAction](https://github.com/keycloak/keycloak/blob/main/server-spi/src/main/java/org/keycloak/models/UserModel.java#L233)  however, since users can define their own required actions with custom aliases, we cannot use a static list here.

File: js/libs/keycloak-js/dist/keycloak.d.ts
Comment (thomasdarimont) [reply js/libs/keycloak-js/dist/keycloak.d.ts]: I added some example values there. That should be enough for users to figure out what they can expect there.

**Review by jonkoops — APPROVED** [review]

**Review by mposolda — APPROVED** [review]

**Comment by mposolda** [comment]
@thomasdarimont @jonkoops Thanks for all the updates and for the review! I hope to merge once tests are OK (which I hope that will be as operator tests failures were probably unrelated to this PR).

**Pull request merged**
